controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -946,118 +946,169 @@ private boolean isSubscriberToNonSubscriberTransition(final ReaderGroupConfigRec
-                    .max(Comparator.comparingLong(StreamCutReferenceRecord::getRecordingTime));
-                    // if retainedSize is less than min size then do not truncate the stream.
-                    if (retainedSize < policy.getRetentionParam()) {
-                        // if retainedSize is less than min size then truncate the stream at max bound.
-                        return maxBound.get().map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor)
-                        .thenApply(StreamCutRecord::getStreamCut))
-                        .orElse(CompletableFuture.completedFuture(null));
-                        } else {
+                .thenCompose(sizeTillLB -> {
+                    long retainedSizeLB = currentSize - sizeTillLB;
+                    // if retainedSize is less than min size then we need to truncate at min or the most recent streamcut
+                    // strictly less than lowerbound. 
+                    if (retainedSizeLB < policy.getRetentionParam()) {
+                        // if no overlap with min then truncate at min
+                        // else truncate at streamcut before lb
+                        return Optional.ofNullable(limits.getValue()).map(x ->
+                                streamMetadataStore.getStreamCutRecord(scope, stream, limits.getValue(), context, executor)
+                                   .thenCompose(limitMin -> {
+                                       return streamMetadataStore.compareStreamCut(scope, stream, limitMin.getStreamCut(), lowerBound, context, executor)
+                                         .thenCompose(compareWithMin -> {
+                                             switch (compareWithMin) {
+                                                 case LessThan: // min less than (behind) lowerbound. truncate at min
+                                                     return CompletableFuture.completedFuture(limitMin.getStreamCut());
+                                                 default:
+                                                     // we cannot have min greater (ahead) than lowerbound as retainedSizeLB < min.
+                                                     // so this is the overlapping case. we need to find streamcut before lowerbound
+                                                     // and truncate at it.
+                                                     // since this is a caught up bound (meaning truncating at subscribers LB will break min policy), 
+                                                     // we cannot force truncate at the max if max overlapped with LB.
+                                                     // so we will find the streamcut from retention set before/lessThan LB (and since it overlaps with min 
+                                                     // so it will definitely be before/lessThan min). 
```

Comment:
We'd never want to truncate at a stream-cut before/lesser-than than min.
The concepts of before/after, behind/ahead are very confusing to grasp here.
It may be good to explain the _terms to describe relative positions of stream-cuts_  ( ex:  `behind/ahead or before/after or greaterThan/lessThan` ) in some method definition and then use just 1 set of terms -- using multiple of these interchangably makes things very confusing.
My guess is :
behind/before/lesserThan == further from the Stream tail
ahead/after/greaterThan == closer to the Stream tail.
So `SC1` is greaterThan `SC2` means `SC1` is closer to the Stream tail than `SC2`

Reply:
It is important to go back to definition of a stream cut. Then it becomes obvious that greater than after and ahead mean same thing. 
a stream cut is an artifact that splits the stream into two halves. So logically each event position will either be in first half or second half or on the stream cut.
Now before/behind etc are events that were written before this stream cut.. ahead/after/greater than are events that were written after this stream cut.